File: producer_consumer.java
import java.util.Scanner;
import java.util.concurrent.Semaphore;

public class producer_consumer {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Get the number of items to produce and consume from the user
        System.out.print("Enter the number of items to produce and consume: ");
        int numItems = scanner.nextInt();

        Semaphore semaphoreProducer = new Semaphore(1);
        Semaphore semaphoreConsumer = new Semaphore(0);

        Producer producer = new Producer(semaphoreProducer, semaphoreConsumer, numItems);
        Consumer consumer = new Consumer(semaphoreConsumer, semaphoreProducer, numItems);

        Thread producerThread = new Thread(producer, "ProducerThread");
        Thread consumerThread = new Thread(consumer, "ConsumerThread");

        producerThread.start();
        consumerThread.start();

        scanner.close(); // Close the scanner after use
    }
}

class Producer implements Runnable {
    Semaphore semaphoreProducer;
    Semaphore semaphoreConsumer;
    int numItems;

    public Producer(Semaphore semaphoreProducer, Semaphore semaphoreConsumer, int numItems) {
        this.semaphoreProducer = semaphoreProducer;
        this.semaphoreConsumer = semaphoreConsumer;
        this.numItems = numItems;
    }

    public void run() {
        for (int i = 1; i <= numItems; i++) {
            try {
                semaphoreProducer.acquire();
                System.out.println("Produced : " + i);
                System.out.println("Consumer is waiting ");

                Thread.sleep(1000);
                semaphoreConsumer.release();

            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}

class Consumer implements Runnable {
    Semaphore semaphoreConsumer;
    Semaphore semaphoreProducer;
    int numItems;

    public Consumer(Semaphore semaphoreConsumer, Semaphore semaphoreProducer, int numItems) {
        this.semaphoreConsumer = semaphoreConsumer;
        this.semaphoreProducer = semaphoreProducer;
        this.numItems = numItems;
    }

    public void run() {
        for (int i = 1; i <= numItems; i++) {
            try {
                semaphoreConsumer.acquire();
                System.out.println("Consumed : " + i);
                System.out.println("Producer is waiting ");

                Thread.sleep(1000);
                semaphoreProducer.release();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}

File: producer_consumer.cpp
#include <iostream>
#include <thread>
#include <semaphore.h> // For semaphore functionality
#include <unistd.h> // For sleep function

// Semaphore declarations
sem_t semaphoreProducer;
sem_t semaphoreConsumer;

// Producer class definition
class Producer {
public:
    // Constructor
    Producer(sem_t &semProducer, sem_t &semConsumer, int numIterations) : semaphoreProducer(semProducer), semaphoreConsumer(semConsumer), iterations(numIterations) {}

    // Overloaded function call operator
    void operator()() {
        // Loop for producing items
        for (int i = 1; i <= iterations; i++) {
            sem_wait(&semaphoreProducer); // Wait for producer semaphore
            std::cout << "Produced: " << i << std::endl; // Print produced item
            std::cout << "Consumer is waiting" << std::endl; // Indicate consumer is waiting
            sleep(1); // Simulate processing time
            sem_post(&semaphoreConsumer); // Signal consumer semaphore
        }
    }

private:
    sem_t &semaphoreProducer; // Reference to producer semaphore
    sem_t &semaphoreConsumer; // Reference to consumer semaphore
    int iterations; // Number of iterations
};

// Consumer class definition
class Consumer {
public:
    // Constructor
    Consumer(sem_t &semConsumer, sem_t &semProducer, int numIterations) : semaphoreConsumer(semConsumer), semaphoreProducer(semProducer), iterations(numIterations) {}

    // Overloaded function call operator
    void operator()() {
        // Loop for consuming items
        for (int i = 1; i <= iterations; i++) {
            sem_wait(&semaphoreConsumer); // Wait for consumer semaphore
            std::cout << "Consumed: " << i << std::endl; // Print consumed item
            std::cout << "Producer is waiting" << std::endl; // Indicate producer is waiting
            sleep(1); // Simulate processing time
            sem_post(&semaphoreProducer); // Signal producer semaphore
        }
    }

private:
    sem_t &semaphoreConsumer; // Reference to consumer semaphore
    sem_t &semaphoreProducer; // Reference to producer semaphore
    int iterations; // Number of iterations
};

int main() {
    int numIterations;
    std::cout << "Enter the number of iterations: ";
    std::cin >> numIterations;

    // Initialize semaphores
    sem_init(&semaphoreProducer, 0, 1);
    sem_init(&semaphoreConsumer, 0, 0);

    // Create instances of Producer and Consumer classes
    Producer producer(semaphoreProducer, semaphoreConsumer, numIterations);
    Consumer consumer(semaphoreConsumer, semaphoreProducer, numIterations);

    // Create threads for producer and consumer
    std::thread producerThread(producer);
    std::thread consumerThread(consumer);

    // Join threads
    producerThread.join();
    consumerThread.join();

    // Destroy semaphores
    sem_destroy(&semaphoreProducer);
    sem_destroy(&semaphoreConsumer);

    return 0;
}
